chat.go

```diff
@@ -61,8 +62,61 @@ type ChatCompletionRequest struct {
 type FunctionDefine struct {
 	Name        string `json:"name"`
 	Description string `json:"description,omitempty"`
-	// it's required in function call
-	Parameters *FunctionParams `json:"parameters"`
+	// ParametersRaw is a JSONSchema object describing the function.
+	// You can pass a raw byte array describing the schema,
+	// or you can generate the array from a JSONSchema object, using another library.
+	ParametersRaw json.RawMessage `json:"-"`
+	// Deprecated: DO NOT USE. Use ParametersRaw instead.
+	Parameters *FunctionParams `json:"-"`
+}
+
+func (fd FunctionDefine) MarshalJSON() ([]byte, error) {
+	type Alias FunctionDefine
+	var parameters json.RawMessage
+	var err error
+
+	if fd.ParametersRaw != nil {
+		parameters = fd.ParametersRaw
+	} else {
+		parameters, err = json.Marshal(fd.Parameters)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return json.Marshal(&struct {
+		*Alias
+		Parameters json.RawMessage `json:"parameters"`
+	}{
+		Alias:      (*Alias)(&fd),
+		Parameters: parameters,
+	})
+}
+
+func (fd *FunctionDefine) UnmarshalJSON(data []byte) error {
+	type Alias FunctionDefine
```

Comment:
How about adding a comment indicating that FunctionDefine type aliases are used to prevent recursion during JSON unmarshalling?